gRPC Recipe Error

Hi everyone,

I continue the issue described here: https://github.com/kivy/python-for-android/issues/1052 . The idea is to create a gRPC recipe for p4a.

I tried to fork the existing recipe, https://github.com/hpsaturn/p4a_grpc_recipe/blob/master/__init__.py. The recipe there was built inside the docker container, and all the dependencies were also installed separately.
In my fork https://github.com/Splato/p4a_grpc_recipe/edit/master/__init__.py I install all dependencies as 'modules' (code lines 129-133) except for the openssl package (grpc installs BoringSSL that fails in p4a). However, I get the following error after compiling 60% of the recipe:

```
[ 60%] Linking CXX shared library libgrpc_plugin_support.so
CMakeFiles/grpc_plugin_support.dir/src/compiler/python_generator.cc.o: In function `PythonGrpcGenerator':
~/.buildozer/android/platform/build-arm64-v8a/build/other_builds/grpcio/arm64-v8a__ndk_target_21/grpcio/grpc/src/compiler/python_generator.cc:829: undefined reference to `google::protobuf::compiler::CodeGenerator::~CodeGenerator()'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [CMakeFiles/grpc_plugin_support.dir/build.make:175: libgrpc_plugin_support.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:894: CMakeFiles/grpc_plugin_support.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```
Then I also tried to include protobuf as dependencies in the grpc recipe (the recipe protobuf_cpp https://github.com/kivy/python-for-android/blob/master/pythonforandroid/recipes/protobuf_cpp/__init__.py). Then I changed the protobuf to 'package' (129) and the path to libraries and executables in cmake (lines 112-115). However, then I got another error:
```
-- Looking for __system_property_get - found
CMake Error at /usr/share/cmake-3.16/Modules/FindProtobuf.cmake:469 (file):
  file STRINGS file "/opt/install/include/google/protobuf/stubs/common.h"
  cannot be read.
Call Stack (most recent call first):
  cmake/protobuf.cmake:58 (find_package)
  CMakeLists.txt:261 (include)


CMake Error at /usr/share/cmake-3.16/Modules/FindProtobuf.cmake:475 (math):
  math cannot parse the expression: " / 1000000": syntax error, unexpected
  exp_DIVIDE (2).
Call Stack (most recent call first):
  cmake/protobuf.cmake:58 (find_package)
  CMakeLists.txt:261 (include)


CMake Error at /usr/share/cmake-3.16/Modules/FindProtobuf.cmake:476 (math):
  math cannot parse the expression: " / 1000 % 1000": syntax error,
  unexpected exp_DIVIDE (2).
Call Stack (most recent call first):
  cmake/protobuf.cmake:58 (find_package)
  CMakeLists.txt:261 (include)


CMake Error at /usr/share/cmake-3.16/Modules/FindProtobuf.cmake:477 (math):
  math cannot parse the expression: " % 1000": syntax error, unexpected
  exp_MOD (2).
Call Stack (most recent call first):
  cmake/protobuf.cmake:58 (find_package)
  CMakeLists.txt:261 (include)


-- Found Protobuf: /opt/install/lib/libprotobuf.a (found version "ERROR.ERROR.ERROR") 
-- Found OpenSSL: ~/.buildozer/android/platform/build-arm64-v8a/build/other_builds/openssl/arm64-v8a__ndk_target_21/openssl1.1/libcrypto.so (found version "1.1.1k")  
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of off64_t
-- Check size of off64_t - done
-- Looking for fseeko
-- Looking for fseeko - found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Renaming
--     ~/.buildozer/android/platform/build-arm64-v8a/build/other_builds/grpcio/arm64-v8a__ndk_target_21/grpcio/grpc/third_party/zlib/zconf.h
-- to 'zconf.h.included' because this file is included with zlib
-- but CMake generates it automatically in the build directory.
-- Configuring incomplete, errors occurred!
See also "~/.buildozer/android/platform/build-arm64-v8a/build/other_builds/grpcio/arm64-v8a__ndk_target_21/grpcio/grpc/build/CMakeFiles/CMakeOutput.log".
See also "~/.buildozer/android/platform/build-arm64-v8a/build/other_builds/grpcio/arm64-v8a__ndk_target_21/grpcio/grpc/build/CMakeFiles/CMakeError.log".
```
So I guess at this point, I need some help with configuring the protobuf installation. Thanks.

After specifying the correct path to libprotobuf.so and libprotoc.so (-DProtobuf_PROTOC_EXECUTABLE,.Dprotobuf_LIBRARIES',' DProtobuf_PROTOC_LIBRARY','-DProtobuf_INCLUDE_DIR',) I removed these errors. Now the recipe is compiled to 99%, but in the end I get the following errors:

```
[ 95%] Linking CXX executable gen_hpack_tables
/home/sergey/.buildozer/android/platform/android-ndk-r19c/toolchains/llvm/prebuilt/linux-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: warning: libpython3.8.so, needed by /home/sergey/Desktop/ld_mediapipe/.buildozer/android/platform/build-arm64-v8a/build/other_builds/protobuf_test/arm64-v8a__ndk_target_24/protobuf_test/src/.libs/libprotobuf.so, not found (try using -rpath or -rpath-link)
/home/sergey/.buildozer/android/platform/android-ndk-r19c/toolchains/llvm/prebuilt/linux-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: warning: libz.so, needed by /home/sergey/Desktop/ld_mediapipe/.buildozer/android/platform/build-arm64-v8a/build/other_builds/protobuf_test/arm64-v8a__ndk_target_24/protobuf_test/src/.libs/libprotobuf.so, not found (try using -rpath or -rpath-link)
/home/sergey/.buildozer/android/platform/android-ndk-r19c/toolchains/llvm/prebuilt/linux-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: warning: libc++_shared.so, needed by /home/sergey/Desktop/ld_mediapipe/.buildozer/android/platform/build-arm64-v8a/build/other_builds/protobuf_test/arm64-v8a__ndk_target_24/protobuf_test/src/.libs/libprotobuf.so, not found (try using -rpath or -rpath-link)
/home/sergey/.buildozer/android/platform/android-ndk-r19c/toolchains/llvm/prebuilt/linux-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: warning: libstdc++.so, needed by /home/sergey/Desktop/ld_mediapipe/.buildozer/android/platform/build-arm64-v8a/build/other_builds/protobuf_test/arm64-v8a__ndk_target_24/protobuf_test/src/.libs/libprotobuf.so, not found (try using -rpath or -rpath-link)
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [CMakeFiles/gen_hpack_tables.dir/build.make:151: gen_hpack_tables] Error 1
make[1]: *** [CMakeFiles/Makefile2:1569: CMakeFiles/gen_hpack_tables.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
I tried already to add these dependencies to LDFLAGS and CXXFLAGS in the environment, specify them in -DCMAKE_INSTALL_PREFIX and -DCMAKE_FIND_ROOT_PATH in the cmake, change -DCMAKE_SHARED_LINKER_FLAGS and -DCMAKE_EXE_LINKER_FLAGS. All these ideas unfortunately don't work. Somehow the libprotobuf.so file compiled with the protobuf recipe cannot find the c++share and python libraries. Perhaps anyone has an idea what else can I try to solve it. Thanks!

--- a/pythonforandroid/recipes/grpcio/__init__.py
+++ b/pythonforandroid/recipes/grpcio/__init__.py
@@ -5,7 +5,11 @@ class GrpcioRecipe(PyProjectRecipe):
     version = '1.64.0'
     url = 'https://files.pythonhosted.org/packages/source/g/grpcio/grpcio-{version}.tar.gz'
     depends = ["setuptools", "librt", "libpthread"]
-    patches = ["comment-getserverbyport-r-args.patch", "remove-android-log-write.patch", "use-ndk-zlib-and-openssl-recipe-include.patch"]
+    patches = [
+        "comment-getserverbyport-r-args.patch",
+        "remove-android-log-write.patch",
+        "use-ndk-zlib-and-openssl-recipe-include.patch"
+    ]
 
     def get_recipe_env(self, arch, **kwargs):
         env = super().get_recipe_env(arch, **kwargs)

--- a/pythonforandroid/recipes/grpcio/__init__.py
+++ b/pythonforandroid/recipes/grpcio/__init__.py
@@ -1,13 +1,7 @@
-import glob
-from logging import info
-
-import sh
-from pythonforandroid.logger import shprint
-from pythonforandroid.recipe import CppCompiledComponentsPythonRecipe
-from pythonforandroid.util import current_directory
+from pythonforandroid.recipe import PyProjectRecipe, Recipe
 
 
-class GrpcioRecipe(CppCompiledComponentsPythonRecipe):
+class GrpcioRecipe(PyProjectRecipe):
     version = '1.64.0'
     url = 'https://files.pythonhosted.org/packages/source/g/grpcio/grpcio-{version}.tar.gz'
     depends = ["setuptools", "librt", "libpthread"]
@@ -15,32 +9,20 @@ class GrpcioRecipe(CppCompiledComponentsPythonRecipe):
 
     def get_recipe_env(self, arch, **kwargs):
         env = super().get_recipe_env(arch, **kwargs)
-        env['NDKPLATFORM'] = "NOTNONE"
-        env['GRPC_PYTHON_BUILD_WITH_CYTHON'] = '1'
+        env["NDKPLATFORM"] = "NOTNONE"
+        env["GRPC_PYTHON_BUILD_SYSTEM_OPENSSL"] = "1"
+        env["GRPC_PYTHON_BUILD_SYSTEM_ZLIB"] = "1"
+        env["ZLIB_INCLUDE"] = self.ctx.ndk.sysroot_include_dir
+        # replace -I with a space
+        openssl_recipe = Recipe.get_recipe('openssl', self.ctx)
+        env["SSL_INCLUDE"] = openssl_recipe.include_flags(arch).strip().replace("-I", "")
         env["CFLAGS"] += " -U__ANDROID_API__"
         env["CFLAGS"] += " -D__ANDROID_API__={}".format(self.ctx.ndk_api)
-
         # turn off c++11 warning error of "invalid suffix on literal"
         env["CFLAGS"] += " -Wno-reserved-user-defined-literal"
-        env['PLATFORM'] = 'android'
+        env["PLATFORM"] = "android"
         env["LDFLAGS"] += " -llog -landroid"
         return env
-
-    def build_compiled_components(self, arch):
-        info('Building compiled components in {}'.format(self.name))
-
-        env = self.get_recipe_env(arch)
-        hostpython = sh.Command(self.hostpython_location)
-        with current_directory(self.get_build_dir(arch.arch)):
-            if self.install_in_hostpython:
-                shprint(hostpython, 'setup.py', 'clean', '--all', _env=env)
-            shprint(hostpython, 'setup.py', self.build_cmd, '-v',
-                    _env=env, *self.setup_extra_args)
-
-            # grpcio creates a build directory and names it `pyb`
-            build_dir = glob.glob('pyb/lib.*')[0]
-            shprint(sh.find, build_dir, '-name', '"*.o"', '-exec',
-                    env['STRIP'], '{}', ';', _env=env)
 
 
 recipe = GrpcioRecipe()

--- a/pythonforandroid/recipes/grpcio/__init__.py
+++ b/pythonforandroid/recipes/grpcio/__init__.py
@@ -26,6 +26,7 @@ class GrpcioRecipe(PyProjectRecipe):
         env["CFLAGS"] += " -Wno-reserved-user-defined-literal"
         env["PLATFORM"] = "android"
         env["LDFLAGS"] += " -llog -landroid"
+        env["LDFLAGS"] += openssl_recipe.link_flags(arch)
         return env
 
 

--- a/pythonforandroid/recipes/grpcio/__init__.py
+++ b/pythonforandroid/recipes/grpcio/__init__.py
@@ -1,0 +1,46 @@
+import glob
+from logging import info
+
+import sh
+from pythonforandroid.logger import shprint
+from pythonforandroid.recipe import CppCompiledComponentsPythonRecipe
+from pythonforandroid.util import current_directory
+
+
+class GrpcioRecipe(CppCompiledComponentsPythonRecipe):
+    version = '1.64.0'
+    url = 'https://files.pythonhosted.org/packages/source/g/grpcio/grpcio-{version}.tar.gz'
+    depends = ["setuptools", "librt", "libpthread"]
+    patches = ["comment-getserverbyport-r-args.patch", "remove-android-log-write.patch"]
+
+    def get_recipe_env(self, arch, **kwargs):
+        env = super().get_recipe_env(arch, **kwargs)
+        env['NDKPLATFORM'] = "NOTNONE"
+        env['GRPC_PYTHON_BUILD_WITH_CYTHON'] = '1'
+        env["CFLAGS"] += " -U__ANDROID_API__"
+        env["CFLAGS"] += " -D__ANDROID_API__={}".format(self.ctx.ndk_api)
+
+        # turn off c++11 warning error of "invalid suffix on literal"
+        env["CFLAGS"] += " -Wno-reserved-user-defined-literal"
+        env['PLATFORM'] = 'android'
+        env["LDFLAGS"] += " -llog -landroid"
+        return env
+
+    def build_compiled_components(self, arch):
+        info('Building compiled components in {}'.format(self.name))
+
+        env = self.get_recipe_env(arch)
+        hostpython = sh.Command(self.hostpython_location)
+        with current_directory(self.get_build_dir(arch.arch)):
+            if self.install_in_hostpython:
+                shprint(hostpython, 'setup.py', 'clean', '--all', _env=env)
+            shprint(hostpython, 'setup.py', self.build_cmd, '-v',
+                    _env=env, *self.setup_extra_args)
+
+            # grpcio creates a build directory and names it `pyb`
+            build_dir = glob.glob('pyb/lib.*')[0]
+            shprint(sh.find, build_dir, '-name', '"*.o"', '-exec',
+                    env['STRIP'], '{}', ';', _env=env)
+
+
+recipe = GrpcioRecipe()

--- a/pythonforandroid/recipes/grpcio/__init__.py
+++ b/pythonforandroid/recipes/grpcio/__init__.py
@@ -5,7 +5,7 @@ class GrpcioRecipe(PyProjectRecipe):
     version = '1.64.0'
     url = 'https://files.pythonhosted.org/packages/source/g/grpcio/grpcio-{version}.tar.gz'
     depends = ["setuptools", "librt", "libpthread"]
-    patches = ["comment-getserverbyport-r-args.patch", "remove-android-log-write.patch"]
+    patches = ["comment-getserverbyport-r-args.patch", "remove-android-log-write.patch", "use-ndk-zlib-and-openssl-recipe-include.patch"]
 
     def get_recipe_env(self, arch, **kwargs):
         env = super().get_recipe_env(arch, **kwargs)

--- a/pythonforandroid/recipes/grpcio/comment-getserverbyport-r-args.patch
+++ b/pythonforandroid/recipes/grpcio/comment-getserverbyport-r-args.patch
@@ -1,0 +1,17 @@
+Index: ares_config.h
+IDEA additional info:
+Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
+<+>UTF-8
+===================================================================
+diff --git a/third_party/cares/config_android/ares_config.h b/third_party/cares/config_android/ares_config.h
+--- a/third_party/cares/config_android/ares_config.h
++++ b/third_party/cares/config_android/ares_config.h	(date 1716777985227)
+@@ -43,7 +43,7 @@
+ #define GETNAMEINFO_TYPE_ARG7 int
+ 
+ /* Specifies the number of arguments to getservbyport_r */
+-#define GETSERVBYPORT_R_ARGS
++//#define GETSERVBYPORT_R_ARGS
+ 
+ /* Define to 1 if you have AF_INET6. */
+ #define HAVE_AF_INET6

--- a/pythonforandroid/recipes/grpcio/remove-android-log-write.patch
+++ b/pythonforandroid/recipes/grpcio/remove-android-log-write.patch
@@ -1,0 +1,36 @@
+Index: log.cc
+IDEA additional info:
+Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
+<+>UTF-8
+===================================================================
+diff --git a/src/core/lib/gpr/android/log.cc b/src/core/lib/gpr/android/log.cc
+--- a/src/core/lib/gpr/android/log.cc
++++ b/src/core/lib/gpr/android/log.cc	(date 1716778822204)
+@@ -30,18 +30,6 @@
+ 
+ #include "src/core/lib/gprpp/crash.h"
+ 
+-static android_LogPriority severity_to_log_priority(gpr_log_severity severity) {
+-  switch (severity) {
+-    case GPR_LOG_SEVERITY_DEBUG:
+-      return ANDROID_LOG_DEBUG;
+-    case GPR_LOG_SEVERITY_INFO:
+-      return ANDROID_LOG_INFO;
+-    case GPR_LOG_SEVERITY_ERROR:
+-      return ANDROID_LOG_ERROR;
+-  }
+-  return ANDROID_LOG_DEFAULT;
+-}
+-
+ void gpr_log(const char* file, int line, gpr_log_severity severity,
+              const char* format, ...) {
+   // Avoid message construction if gpr_log_message won't log
+@@ -70,8 +58,6 @@
+ 
+   asprintf(&output, "%s:%d] %s", display_file, args->line, args->message);
+ 
+-  __android_log_write(severity_to_log_priority(args->severity), "GRPC", output);
+-
+   // allocated by asprintf => use free, not gpr_free
+   free(output);
+ }

--- a/pythonforandroid/recipes/grpcio/use-ndk-zlib-and-openssl-recipe-include.patch
+++ b/pythonforandroid/recipes/grpcio/use-ndk-zlib-and-openssl-recipe-include.patch
@@ -1,0 +1,16 @@
+--- a/setup.py	2024-05-31 11:20:56.824695569 +0100
++++ b/setup.py	2024-05-31 23:13:40.324392463 +0100
+@@ -299,11 +299,11 @@
+         lambda x: "third_party/boringssl" not in x, CORE_C_FILES
+     )
+     CORE_C_FILES = filter(lambda x: "src/boringssl" not in x, CORE_C_FILES)
+-    SSL_INCLUDE = (os.path.join("/usr", "include", "openssl"),)
++    SSL_INCLUDE = tuple(os.environ["SSL_INCLUDE"].split(" "))
+ 
+ if BUILD_WITH_SYSTEM_ZLIB:
+     CORE_C_FILES = filter(lambda x: "third_party/zlib" not in x, CORE_C_FILES)
+-    ZLIB_INCLUDE = (os.path.join("/usr", "include"),)
++    ZLIB_INCLUDE = tuple(os.environ["ZLIB_INCLUDE"].split(" "))
+ 
+ if BUILD_WITH_SYSTEM_CARES:
+     CORE_C_FILES = filter(lambda x: "third_party/cares" not in x, CORE_C_FILES)

